core/src/main/java/io/ultrabrew/metrics/data/BasicHistogramAggregator.java

```diff
@@ -0,0 +1,214 @@
+// Copyright 2019, Oath Inc.
+// Licensed under the terms of the Apache License 2.0 license. See LICENSE file in Ultrabrew Metrics
+// for terms.
+
+package io.ultrabrew.metrics.data;
+
+import static io.ultrabrew.metrics.Metric.DEFAULT_CARDINALITY;
+import static io.ultrabrew.metrics.Metric.DEFAULT_MAX_CARDINALITY;
+
+import io.ultrabrew.metrics.Histogram;
+import io.ultrabrew.metrics.util.DistributionBucket;
+import java.util.Arrays;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import sun.misc.Unsafe;
+
+public class BasicHistogramAggregator extends ConcurrentIntTable implements Aggregator {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(BasicHistogramAggregator.class);
+  private static final String[] AGG_FIELDS = {"count", "sum", "min", "max"};
+  private static final long[] IDENTITY = {0L, 0L, Long.MAX_VALUE, Long.MIN_VALUE};
+
+  private final String metricId;
+  private final DistributionBucket buckets;
+  private final int bucketCount;
+
+  public BasicHistogramAggregator(Histogram histogram) {
+    this(histogram.id, histogram.bucket, histogram.cardinality, histogram.maxCardinality);
+  }
+
+  public BasicHistogramAggregator(final String metricId, final DistributionBucket buckets) {
+    this(metricId, buckets, DEFAULT_CARDINALITY, DEFAULT_MAX_CARDINALITY);
+  }
+
+  public BasicHistogramAggregator(final String metricId, final DistributionBucket buckets, final int capacity) {
+    this(metricId, buckets, capacity, DEFAULT_MAX_CARDINALITY);
+  }
+
+  public BasicHistogramAggregator(final String metricId, final DistributionBucket buckets, final int capacity, final int maxCapacity) {
+    super(AGG_FIELDS.length + buckets.getCount(), capacity, maxCapacity, IDENTITY);
+    this.metricId = metricId;
+    this.buckets = buckets;
+    this.bucketCount = buckets.getCount();
+  }
+
+  @Override
+  public void apply(String[] tags, long latency, long timestamp) {
+    apply(tags, (int) latency, timestamp);
+  }
+
+  @Override
+  protected void combine(int[] table, long baseOffset, long value) {
+    addToAggField(table, baseOffset, 0, 1);
+    addToAggField(table, baseOffset, 1, value);
+    minAggField(table, baseOffset, 2, value);
+    maxAggField(table, baseOffset, 3, value);
+
+    //Increments the bucket counter by 1 responding to the given value
+    int bucketIndex = buckets.getBucketIndex(value);
+    addToDataField(table, baseOffset, AGG_FIELDS.length + bucketIndex, 1);
+  }
+
+  @Override
+  public Cursor cursor() {
+    return newCursor(false);
+  }
+
+  @Override
+  public Cursor sortedCursor() {
+    return newCursor(true);
+  }
+
+  private Cursor newCursor(boolean sorted) {
+    String[] fields = buildFields();
+    Type[] types = buildTypes(fields.length);
```

Comment:
This should be cached in the aggregator instead of being recreated every time a new cursor is created.